src/libraries/System.Linq/tests/RangeIBinaryIntegerTests.cs

```diff
@@ -0,0 +1,1089 @@
+            { ulong.MaxValue - int.MaxValue + 2, -1 },
+            { ulong.MaxValue - int.MaxValue + 2, int.MinValue },
+
+            { ulong.MaxValue - int.MaxValue / 2, int.MaxValue / 2 + 2 },
+            { ulong.MaxValue - int.MaxValue / 2, int.MaxValue },
+            { ulong.MaxValue - int.MaxValue / 2, -1 },
+            { ulong.MaxValue - int.MaxValue / 2, int.MinValue },
+
+            { ulong.MaxValue - 1, 3 },
+            { ulong.MaxValue - 1, int.MaxValue / 4 },
+            { ulong.MaxValue - 1, int.MaxValue / 2 },
+            { ulong.MaxValue - 1, int.MaxValue },
+            { ulong.MaxValue - 1, -1 },
+            { ulong.MaxValue - 1, int.MinValue },
+
+            { ulong.MaxValue, 2 },
+            { ulong.MaxValue, int.MaxValue / 4 },
+            { ulong.MaxValue, int.MaxValue / 2 },
+            { ulong.MaxValue, int.MaxValue },
+            { ulong.MaxValue, -1 },
+            { ulong.MaxValue, int.MinValue },
+        };
+    }
+
+    public class RangeLongTests : RangeIBinaryIntegerTests<long>
+    {
+        public static TheoryData<long, int> StartCountCorrectData { get; } = new TheoryData<long, int>()
+        {
+            { long.MinValue, 15 },
+            { long.MinValue, 1 },
+            { long.MinValue, 0 },
+
+
+            { -1, 15 },
+            { -1, 1 },
+            { -1, 0 },
+
+            { long.MaxValue - 14, 15 },
+            { long.MaxValue, 1 },
+            { long.MaxValue, 0 },
+        };
+
+        public static TheoryData<long, int> StartCountIncorrectData { get; } = new TheoryData<long, int>()
+        {
+            { long.MinValue, -1 },
+            { long.MinValue, int.MinValue },
+
+            { long.MinValue + 1, -1 },
+            { long.MinValue + 1, int.MinValue },
+
+            { long.MinValue / 2, -1 },
+            { long.MinValue / 2, int.MinValue },
+
+            { 0, -1 },
+            { 0, int.MinValue },
+
+            { 2, -1 },
+            { 2, int.MinValue },
+
+            { long.MaxValue / 2, -1 },
+            { long.MaxValue / 2, int.MinValue },
+
+            { long.MaxValue - int.MaxValue + 2, int.MaxValue },
+            { long.MaxValue - int.MaxValue + 2, -1 },
+            { long.MaxValue - int.MaxValue + 2, int.MinValue },
+
+            { long.MaxValue - int.MaxValue / 2, int.MaxValue / 2 + 2 },
+            { long.MaxValue - int.MaxValue / 2, int.MaxValue },
+            { long.MaxValue - int.MaxValue / 2, -1 },
+            { long.MaxValue - int.MaxValue / 2, int.MinValue },
+
+            { long.MaxValue - 1, 3 },
+            { long.MaxValue - 1, int.MaxValue },
+            { long.MaxValue - 1, -1 },
+            { long.MaxValue - 1, int.MinValue },
+
+            { long.MaxValue, 2 },
+            { long.MaxValue, int.MaxValue },
+            { long.MaxValue, -1 },
+            { long.MaxValue, int.MinValue },
+        };
+    }
+
+    public class RangeBigIntegerTests : RangeIBinaryIntegerTests<BigInteger>
```

Comment:
It would be interesting to see a few nonstandard numeric types that stress test the implementation, e.g. a modular arithmetic type. If for example I define a type representing all numbers modulo 7, what should be the behaviour of the new method when I do `Range(Modulo7.Zero, 8)`?